File: frontend/src/types/invoice.ts
export type InvoiceStatus = 'PENDING' | 'PROCESSING' | 'VERIFIED' | 'REJECTED' | 'QUEUED' | 'FAILED' | 'REVIEW_REQUIRED';
export type SourceType = 'LLM' | 'E_INVOICE' | 'MANUAL';
export type LlmProvider = 'GEMINI' | 'GPT' | 'CLAUDE';
export type Currency = 'TRY' | 'USD' | 'EUR' | 'GBP';
export type UnitType = 'ADET' | 'KG' | 'LT' | 'M' | 'M2' | 'M3' | 'PAKET' | 'KUTU' | 'SAAT' | 'GUN' | 'TON';

export interface InvoiceItem {
  id?: string;
  lineNumber: number;
  description: string;
  quantity: number;
  unit: UnitType;
  unitPrice: number;
  taxRate: number;
  taxAmount: number;
  subtotal: number;
  totalAmount: number;
  productCode?: string;
  barcode?: string;
}

export interface InvoiceListItem {
  id: string;
  invoiceNumber: string;
  invoiceDate: string;
  dueDate?: string;
  supplierName: string;
  totalAmount: number;
  currency: Currency;
  status: InvoiceStatus;
  sourceType: SourceType;
  categoryName?: string;
  itemCount: number;
  createdByUserName: string;
  createdAt: string;
}

export interface InvoiceDetail {
  id: string;
  invoiceNumber: string;
  invoiceDate: string;
  dueDate?: string;
  supplierName: string;
  supplierTaxNumber?: string;
  supplierTaxOffice?: string;
  supplierAddress?: string;
  supplierPhone?: string;
  supplierEmail?: string;
  buyerName?: string;
  buyerTaxNumber?: string;
  subtotal: number;
  taxAmount: number;
  totalAmount: number;
  currency: Currency;
  exchangeRate: number;
  status: InvoiceStatus;
  sourceType: SourceType;
  llmProvider?: LlmProvider;
  confidenceScore?: number;
  validationIssues?: ValidationIssue[];
  categoryId?: string;
  categoryName?: string;
  notes?: string;
  rejectionReason?: string;
  createdByUserId: string;
  createdByUserName: string;
  verifiedByUserId?: string;
  verifiedByUserName?: string;
  verifiedAt?: string;
  rejectedAt?: string;
  createdAt: string;
  updatedAt: string;
  items: InvoiceItem[];
}

export interface CreateInvoiceRequest {
  invoiceNumber: string;
  invoiceDate: string;
  dueDate?: string;
  supplierName: string;
  supplierTaxNumber?: string;
  supplierTaxOffice?: string;
  supplierAddress?: string;
  supplierPhone?: string;
  supplierEmail?: string;
  currency: Currency;
  exchangeRate?: number;
  categoryId?: string;
  notes?: string;
  items: CreateInvoiceItemRequest[];
}

export interface CreateInvoiceItemRequest {
  id?: string;
  description: string;
  quantity: number;
  unit: UnitType;
  unitPrice: number;
  taxRate: number;
  productCode?: string;
}

// Update UpdateInvoiceRequest to include corrections
export interface UpdateInvoiceRequest extends CreateInvoiceRequest {
  extractionCorrections?: ExtractionCorrection[];
}

export interface VerifyInvoiceRequest { notes?: string; }
export interface RejectInvoiceRequest { rejectionReason: string; }

export interface PaginatedResponse<T> {
  content: T[];
  page: { size: number; number: number; totalElements: number; totalPages: number; };
}

export interface InvoiceListParams {
  page?: number;
  size?: number;
  sort?: string;
  status?: string; // CSV
  categoryId?: string; // CSV
  dateFrom?: string;
  dateTo?: string;
  amountMin?: number;
  amountMax?: number;
  currency?: string; // CSV
  sourceType?: string; // CSV
  llmProvider?: string; // CSV
  confidenceMin?: number;
  confidenceMax?: number;
  supplierName?: string; // CSV
  search?: string;
  createdByUserId?: string;
  createdFrom?: string;
  createdTo?: string;
}

export interface BatchFileStatus {
  id?: string;
  fileName: string;
  status: 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'FAILED';
  invoiceId?: string;
  error?: string;
}

export interface BatchStatusResponse {
  batchId: string;
  totalFiles: number;
  completedFiles: number;
  failedFiles: number;
  status: 'PROCESSING' | 'COMPLETED' | 'PARTIALLY_COMPLETED' | 'FAILED';
  files: BatchFileStatus[];
}

export interface FilterOptionsResponse {
  statuses: { value: InvoiceStatus; label: string }[];
  categories: { id: string; name: string; color?: string }[];
  currencies: Currency[];
  sourceTypes: SourceType[];
  llmProviders: LlmProvider[];
  amountRange: { min: number; max: number };
  dateRange: { min: string; max: string };
  confidenceRange: { min: number; max: number };
}

export interface ValidationIssue {
  field: string;
  issue: string;
  severity: 'CRITICAL' | 'WARNING' | 'INFO';
}

export interface ExtractionCorrection {
  field: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  originalValue: any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  correctedValue: any;
}

export interface ValidationResult {
  confidenceScore: number;
  validationIssues: ValidationIssue[];
}

export interface InvoiceData {
  invoiceNumber: string;
  invoiceDate: string;
  dueDate?: string;
  supplierName: string;
  supplierTaxNumber?: string;
  supplierAddress?: string;
  buyerName?: string;
  buyerTaxNumber?: string;
  items: InvoiceItem[];
  subtotal: number;
  taxAmount: number;
  totalAmount: number;
  currency: Currency;
  notes?: string;
}

File: backend/src/main/java/com/faturaocr/domain/invoice/entity/Invoice.java
package com.faturaocr.domain.invoice.entity;

import com.faturaocr.domain.invoice.valueobject.Currency;
import com.faturaocr.domain.invoice.valueobject.ExtractionCorrection;
import com.faturaocr.domain.invoice.valueobject.InvoiceStatus;
import com.faturaocr.domain.invoice.valueobject.LlmProvider;
import com.faturaocr.domain.invoice.valueobject.SourceType;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Getter
@Setter
public class Invoice {
    private UUID id;
    private UUID companyId;
    private UUID categoryId;
    private UUID createdByUserId;
    private UUID verifiedByUserId;
    private UUID batchId;
    private String correlationId;

    // Identification
    private String invoiceNumber;
    private LocalDate invoiceDate;
    private LocalDate dueDate;

    // Supplier
    private String supplierName;
    private String supplierTaxNumber;
    private String supplierTaxOffice;
    private String supplierAddress;
    private String supplierPhone;
    private String supplierEmail;

    // Financial
    private BigDecimal subtotal;
    private BigDecimal taxAmount;
    private BigDecimal totalAmount;
    private Currency currency;
    private BigDecimal exchangeRate;

    // Status
    private InvoiceStatus status;

    // Source
    private SourceType sourceType;
    private LlmProvider llmProvider;
    private BigDecimal confidenceScore;
    private Integer processingDurationMs;

    // File
    private String originalFilePath;
    private String storedFilePath;
    private String originalFileName;
    private String fileHash;
    private Integer originalFileSize;
    private String originalFileType;

    // E-Invoice
    private String eInvoiceUuid;
    private String eInvoiceEttn;

    private List<ExtractionCorrection> extractionCorrections;

    // Notes
    private String notes;
    private String rejectionReason;

    // Timestamps
    private LocalDateTime verifiedAt;
    private LocalDateTime rejectedAt;
    private boolean isDeleted;
    private LocalDateTime deletedAt;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    // Items
    private List<InvoiceItem> items = new ArrayList<>();

    public Invoice() {
        this.id = UUID.randomUUID();
        this.status = InvoiceStatus.PENDING;
        this.sourceType = SourceType.MANUAL;
        this.currency = Currency.TRY;
        this.exchangeRate = BigDecimal.ONE;
        this.subtotal = BigDecimal.ZERO;
        this.taxAmount = BigDecimal.ZERO;
        this.totalAmount = BigDecimal.ZERO;
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
        this.isDeleted = false;
    }

    public void addItem(InvoiceItem item) {
        item.setInvoiceId(this.id);
        this.items.add(item);
    }

    public void removeItem(InvoiceItem item) {
        this.items.remove(item);
    }

    public void calculateTotals() {
        this.subtotal = BigDecimal.ZERO;
        this.taxAmount = BigDecimal.ZERO;
        this.totalAmount = BigDecimal.ZERO;

        for (InvoiceItem item : items) {
            this.subtotal = this.subtotal.add(item.getSubtotal());
            this.taxAmount = this.taxAmount.add(item.getTaxAmount());
            this.totalAmount = this.totalAmount.add(item.getTotalAmount());
        }
    }
}
